src/devices/bus/nabupc/adapter.cpp

```diff
@@ -0,0 +1,484 @@
+// license:BSD-3-Clause
+// copyright-holders:Brian Johnson
+/*******************************************************************
+ *
+ * NABU PC - Network Adapter Settop Box
+ *
+ *******************************************************************/
+
+#include "emu.h"
+#include "adapter.h"
+
+#include "emuopts.h"
+#include "unzip.h"
+
+#define VERBOSE 0
+#include "logmacro.h"
+
+
+//**************************************************************************
+//  NABU PC NETWORK ADAPTER DEVICE
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(NABUPC_NETWORK_ADAPTER, bus::nabupc::network_adapter, "nabu_net_adapter", "NABU Network Adapter")
+
+namespace bus::nabupc {
+
+//**************************************************************************
+//  SEGMENT FILE LOADING
+//**************************************************************************
+
+// Load segment file from disk
+std::error_condition network_adapter::segment_file::read_archive(util::core_file &stream, uint32_t segment_id)
+{
+	segment_id &= 0xFFFFFF;
+
+	util::core_file::ptr proxy;
+	std::error_condition err = util::core_file::open_proxy(stream, proxy);
+	if (err)
+		return err;
+
+	util::archive_file::ptr zipfile;
+	err = util::archive_file::open_zip(std::move(proxy), zipfile);
+	if (err)
+		return err;
+
+	std::string segment_filename = util::string_format("%06d.nabu", segment_id);
+
+	if (zipfile->search(segment_filename, false) < 0)
+		return std::errc::no_such_file_or_directory;
+
+	// determine the uncompressed length
+	uint64_t uncompressed_length_uint64 = zipfile->current_uncompressed_length();
+	size_t uncompressed_length = (size_t)uncompressed_length_uint64;
+	if (uncompressed_length != uncompressed_length_uint64)
+		return std::errc::not_enough_memory;
+
+	// prepare a buffer for the segment file
+	std::unique_ptr<char[]> segment_buffer(new (std::nothrow) char[uncompressed_length]);
+	if (!segment_buffer)
+		return std::errc::not_enough_memory;
+
+	err = zipfile->decompress(&segment_buffer[0], uncompressed_length);
+	if (err)
+		return err;
+
+	m_segment_id = segment_id;
+	return parse_segment(&segment_buffer[0], uncompressed_length);
+}
+
+std::error_condition network_adapter::segment_file::parse_segment(char *data, size_t length)
+{
+	util::core_file::ptr fd;
+	pak current;
+	uint64_t offset = 0;
+	size_t actual;
+	uint16_t crc;
+	uint8_t npak = 0;
+	std::error_condition err;
+
+	pak_list.clear();
+	err = util::core_file::open_ram(data, length, OPEN_FLAG_READ, fd);
+	if (err)
+		return err;
+
+	memset(current.data, 0, 991);
+	current.segment_id[0] = (m_segment_id & 0xFF0000) >> 16;
+	current.segment_id[1] = (m_segment_id & 0x00FF00) >> 8;
+	current.segment_id[2] = (m_segment_id & 0xF000FF);
```

Comment:
Could you please check this file for remaining uppercase hex digits and make them lowercase?